### PR TITLE
UCP/PROTO/RNDV: Add 3-stage pipeline put new protocol

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Better RNDV performance with new protocols when GPU-direct is slow or does not exist

### CODE REVIEW

**Comment by yosefe** [comment]
@brminich @dmitrygx can you pls take a look?

File: src/ucp/rndv/proto_rndv.c [hunk 1]
```diff
@@ -534,42 +536,89 @@ ucp_proto_rndv_send_start(ucp_worker_h worker, ucp_request_t *req,
...
+    } else {
+        /* Partial RTR, its "offset" and "size" fields specify part to send */
+        status = ucp_proto_rndv_frag_request_alloc(worker, req, &freq);
+        if (status != UCS_OK) {
+            goto err_request_fail;
+        }
+
+        /* When this fragment is completed, count total size and complete the
+           super request if needed */
+        freq->flags  |= UCP_REQUEST_FLAG_CALLBACK;
```
Comment (dmitrygx) [inline src/ucp/rndv/proto_rndv.c:599]: `ucp_request_set_callback(freq, send.cb, ucp_proto_rndv_send_complete_one, NULL)`

File: src/ucp/rndv/rndv_put.c [hunk 2]
```diff
@@ -310,3 +358,115 @@ static ucp_proto_t ucp_rndv_put_zcopy_proto = {
     },
 };
 UCP_PROTO_REGISTER(&ucp_rndv_put_zcopy_proto);
+
+
+static void ucp_proto_rndv_put_mtype_pack_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
```
Comment (dmitrygx) [inline src/ucp/rndv/rndv_put.c:365]: not alinged by `=`

File: src/ucp/rndv/rndv_put.c [hunk 3]
```diff
@@ -310,3 +358,115 @@ static ucp_proto_t ucp_rndv_put_zcopy_proto = {
...
+
+    ucp_proto_rndv_mtype_next_iov(req, &rpriv->bulk, lpriv, next_iter, &iov);
+    return ucp_proto_rndv_put_common_send(req, lpriv, &iov,
+                                          &req->send.state.uct_comp);
+}
+
+static ucs_status_t
+ucp_proto_rndv_put_mtype_send_progress(uct_pending_req_t *uct_req)
+{
+    ucp_request_t *req = ucs_container_of(uct_req, ucp_request_t, send.uct);
```
Comment (dmitrygx) [inline src/ucp/rndv/rndv_put.c:389]: not aligned by `=`

File: src/ucp/rndv/rndv_rtr.c [hunk 4]
```diff
@@ -9,15 +9,20 @@
...
 
 #include <ucp/proto/proto_single.inl>
 
 
 /**
  * RTR protocol callback, which is called when all incoming data is filled to
  * the request send buffer.
+ *
+ * @param in_buffer Whether data is already in user buffer (dt_iter) or in
+ *  in the buffer we published as remote address.
```
Comment (dmitrygx) [inline src/ucp/rndv/rndv_rtr.c:22]: remove whitespace at the beginning of the line

File: src/ucp/rndv/rndv_rtr.c [hunk 5]
```diff
@@ -219,6 +234,144 @@ static ucp_proto_t ucp_rndv_rtr_proto = {
...
+
+    return UCS_OK;
+}
+
+static ucp_proto_t ucp_rndv_rtr_mtype_proto = {
+    .name       = "rndv/rtr/mtype",
+    .flags      = 0,
+    .init       = ucp_proto_rndv_rtr_mtype_init,
+    .config_str = ucp_proto_rndv_ctrl_config_str,
+    .progress   = {ucp_proto_rndv_rtr_mtype_progress},
```
Comment (dmitrygx) [inline src/ucp/rndv/rndv_rtr.c:371]: no need for trailing comma

File: src/ucp/rndv/rndv_put.c [hunk 6]
```diff
@@ -310,3 +358,115 @@ static ucp_proto_t ucp_rndv_put_zcopy_proto = {
...
+            return UCS_OK;
+        }
+
+        ucp_proto_rndv_put_common_request_init(req);
+        req->flags |= UCP_REQUEST_FLAG_PROTO_INITIALIZED;
+
+        ucp_proto_rndv_mtype_copy(req, uct_ep_get_zcopy,
+                                  ucp_proto_rndv_put_mtype_pack_completion,
+                                  "in from");
+        return UCS_OK;
```
Comment (brminich) [inline src/ucp/rndv/rndv_put.c:406]: why do we return here?

Comment (yosefe) [reply src/ucp/rndv/rndv_put.c:406]: we started the memory type copy operation and will resume sending when this copy is calling its completion callback

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/rndv/rndv_put.c [hunk 7]
```diff
@@ -18,7 +18,11 @@
 
 
 enum {
-    UCP_PROTO_RNDV_PUT_STAGE_SEND = UCP_PROTO_STAGE_START,
+    /* Initial stage for put zcopy is copy the data to the fragment */
```
Comment (yosefe) [inline src/ucp/rndv/rndv_put.c:21]: need to fix this comment

**Review by dmitrygx — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Pull request merged**